File: wanco/src/compile/cr_v2/stackmap/stackmap.rs
/*
ref to https://llvm.org/docs/StackMaps.html#stack-map-format

Header {
  uint8  : Stack Map Version (current version is 3)
  uint8  : Reserved (expected to be 0)
  uint16 : Reserved (expected to be 0)
}
uint32 : NumFunctions
uint32 : NumConstants
uint32 : NumRecords
StkSizeRecord[NumFunctions] {
  uint64 : Function Address
  uint64 : Stack Size (or UINT64_MAX if not statically known)
  uint64 : Record Count
}
Constants[NumConstants] {
  uint64 : LargeConstant
}
StkMapRecord[NumRecords] {
  uint64 : PatchPoint ID
  uint32 : Instruction Offset
  uint16 : Reserved (record flags)
  uint16 : NumLocations
  Location[NumLocations] {
    uint8  : Register | Direct | Indirect | Constant | ConstantIndex
    uint8  : Reserved (expected to be 0)
    uint16 : Location Size
    uint16 : Dwarf RegNum
    uint16 : Reserved (expected to be 0)
    int32  : Offset or SmallConstant
  }
  uint32 : Padding (only if required to align to 8 byte)
  uint16 : Padding
  uint16 : NumLiveOuts
  LiveOuts[NumLiveOuts]
    uint16 : Dwarf RegNum
    uint8  : Reserved
    uint8  : Size in Bytes
  }
  uint32 : Padding (only if required to align to 8 byte)
}
*/

use crate::compile::cr_v2::stackmap::regs::AsStr;

use super::regs::Reg;
use anyhow::{anyhow, Result};
use nom::{
    error::{make_error, ErrorKind},
    multi::count,
    number::complete::{le_i32, le_u16, le_u32, le_u64, le_u8},
    sequence::Tuple,
    IResult,
};

pub fn parse(input: &[u8]) -> Result<Stackmap> {
    // https://stackoverflow.com/questions/55184864/nom-parser-borrow-checker-issue
    let map = parse_stackmap(input).map_err(|e| anyhow!(e.to_string()))?.1;
    Ok(map)
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct Stackmap {
    pub header: Header,
    pub stack_size_records: Box<[StackSizeRecord]>,
    pub constants: Box<[u64]>,
    pub stackmap_records: Box<[StackMapRecord]>,
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct Header {
    pub version: u8,
    pub reserved0: u8,
    pub reserved1: u16,
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct StackSizeRecord {
    pub function_addr: u64,
    pub stack_size: u64,
    pub record_count: u64,
}
#[derive(Debug, Clone, Eq, PartialEq)]
pub struct StackMapRecord {
    pub patchpoint_id: u64,
    pub inst_offset: u32,
    pub locations: Box<[Location]>,
    pub live_outs: Box<[LiveOut]>,
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum LocationValue {
    /// Value is in a register,
    Register { reg: Reg },

    /// Frame index value.
    Direct { reg: Reg, offset: i32 },

    /// Spilled value.
    Indirect { reg: Reg, offset: i32 },

    /// Small constant.
    Constant { value: u32 },

    /// Large constant.
    ConstIndex { index: u32 },
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct Location {
    pub value: LocationValue,
    pub size: u16,
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct LiveOut {
    pub reg: Reg,
    pub size_in_bytes: u8,
}

/*
Header {
  ...
}
uint32 : NumFunctions
uint32 : NumConstants
uint32 : NumRecords
StkSizeRecord[NumFunctions] {
  ...
}
Constants[NumConstants] {
  ...
}
StkMapRecord[NumRecords] {
  ...
  uint32 : Padding (only if required to align to 8 byte)
}
*/
pub fn parse_stackmap(input: &[u8]) -> IResult<&[u8], Stackmap> {
    let (input, header) = parse_header(input)?;
    let (input, (num_functions, num_constants, num_records)) =
        (le_u32, le_u32, le_u32).parse(input)?;
    let (input, stack_size_records) =
        count(parse_stack_size_record, num_functions as usize)(input)?;
    let (input, constants) = count(le_u64, num_constants as usize)(input)?;
    let (input, stackmap_records) = count(parse_stackmap_record, num_records as usize)(input)?;

    Ok((
        input,
        Stackmap {
            header,
            stack_size_records: stack_size_records.into_boxed_slice(),
            constants: constants.into_boxed_slice(),
            stackmap_records: stackmap_records.into_boxed_slice(),
        },
    ))
}

fn parse_header(input: &[u8]) -> IResult<&[u8], Header> {
    let (input, version) = le_u8(input)?;
    let (input, reserved0) = le_u8(input)?;
    let (input, reserved1) = le_u16(input)?;
    Ok((
        input,
        Header {
            version,
            reserved0,
            reserved1,
        },
    ))
}

fn parse_stack_size_record(input: &[u8]) -> IResult<&[u8], StackSizeRecord> {
    let (input, function_addr) = le_u64(input)?;
    let (input, stack_size) = le_u64(input)?;
    let (input, record_count) = le_u64(input)?;
    Ok((
        input,
        StackSizeRecord {
            function_addr,
            stack_size,
            record_count,
        },
    ))
}

/*
StkMapRecord[NumRecords] {
  uint64 : PatchPoint ID
  uint32 : Instruction Offset
  uint16 : Reserved (record flags)
  uint16 : NumLocations
  Location[NumLocations] {
    ...
  }
  uint32 : Padding (only if required to align to 8 byte)
  uint16 : Padding
  uint16 : NumLiveOuts
  LiveOuts[NumLiveOuts]
    ...
  }
  uint32 : Padding (only if required to align to 8 byte)
}
*/
fn parse_stackmap_record(input: &[u8]) -> IResult<&[u8], StackMapRecord> {
    let (input, patchpoint_id) = le_u64(input)?;
    let (input, inst_offset) = le_u32(input)?;
    let (input, _) = le_u16(input)?; // reserved
    let (input, num_locations) = le_u16(input)?;
    let (input, locations) = count(parse_location, num_locations as usize)(input)?;
    // padding (only if required to align to 8 byte)
    let input = if input.as_ptr() as u64 % 8 != 0 {
        le_u32(input)?.0
    } else {
        input
    };
    let (input, _) = le_u16(input)?; // padding
    let (input, num_live_outs) = le_u16(input)?;
    let (input, live_outs) = count(parse_live_out, num_live_outs as usize)(input)?;
    // padding (only if required to align to 8 byte)
    let input = if input.as_ptr() as u64 % 8 != 0 {
        le_u32(input)?.0
    } else {
        input
    };

    Ok((
        input,
        StackMapRecord {
            patchpoint_id,
            inst_offset,
            locations: locations.into_boxed_slice(),
            live_outs: live_outs.into_boxed_slice(),
        },
    ))
}

/*
Location[NumLocations] {
  uint8  : Register | Direct | Indirect | Constant | ConstantIndex
  uint8  : Reserved (expected to be 0)
  uint16 : Location Size
  uint16 : Dwarf RegNum
  uint16 : Reserved (expected to be 0)
  int32  : Offset or SmallConstant
}
*/
fn parse_location(input: &[u8]) -> IResult<&[u8], Location> {
    let (input, kind) = le_u8(input)?;
    let (input, _) = le_u8(input)?; // reserved
    let (input, size) = le_u16(input)?;
    let (input, regnum) = le_u16(input)?;
    let (input, _) = le_u16(input)?; // reserved
    let (input, offset) = le_i32(input)?;

    let value = match kind {
        // Register.
        1 => LocationValue::Register {
            reg: Reg::from(regnum),
        },
        // Direct.
        2 => LocationValue::Direct {
            reg: Reg::from(regnum),
            offset,
        },
        // Indirect.
        3 => LocationValue::Indirect {
            reg: Reg::from(regnum),
            offset,
        },
        // Constant.
        4 => LocationValue::Constant {
            value: offset as u32,
        },
        // ConstIndex.
        5 => LocationValue::ConstIndex {
            index: offset as u32,
        },
        _ => {
            return Err(nom::Err::Error(make_error(input, ErrorKind::Tag)));
        }
    };

    Ok((input, Location { value, size }))
}

fn parse_live_out(input: &[u8]) -> IResult<&[u8], LiveOut> {
    let (input, regnum) = le_u16(input)?;
    let (input, _) = le_u8(input)?;
    let (input, size_in_bytes) = le_u8(input)?;
    Ok((
        input,
        LiveOut {
            reg: Reg::from(regnum),
            size_in_bytes,
        },
    ))
}

pub fn prettyprint(map: &Stackmap) {
    println!("Version: {}", map.header.version);
    println!("NumStackmapRecords: {}", map.stackmap_records.len());
    println!("NumConstants: {}", map.constants.len());
    println!("NumStackSizeRecords: {}", map.stack_size_records.len());
    for stackmap_record in map.stackmap_records.iter() {
        println!("Patchpoint ID: {}", stackmap_record.patchpoint_id);
        println!("- Instruction Offset: {}", stackmap_record.inst_offset);
        for location in stackmap_record.locations.iter() {
            match location.value {
                LocationValue::Register { reg } => {
                    println!(
                        "  - value: {} (size: {}, loc=reg)",
                        reg.as_str(),
                        location.size
                    );
                }
                LocationValue::Direct { reg, offset } => {
                    println!(
                        "  - value: {} + {} (size: {}, loc=direct)",
                        reg.as_str(),
                        offset,
                        location.size
                    );
                }
                LocationValue::Indirect { reg, offset } => {
                    println!(
                        "  - value: [{} {:+}] (size: {}, loc=indirect)",
                        reg.as_str(),
                        offset,
                        location.size
                    );
                }
                LocationValue::Constant { value } => {
                    println!("  - value: {} (size: {}, loc=const)", value, location.size);
                }
                LocationValue::ConstIndex { index } => {
                    println!(
                        "  - value: Constants[{}] (size: {}, loc=const_idx)",
                        index, location.size
                    );
                }
            }
        }
        for live_out in stackmap_record.live_outs.iter() {
            println!(
                "Live Out: {:?}, Size: {}",
                live_out.reg, live_out.size_in_bytes
            );
        }
    }
}

File: lib-rt/v2/stackmap.cc
#include "stackmap.h"
#include "wanco.h"
#include "x86_64.h"
#include <cstdint>
#include <elf.h>
#include <fstream>
#include <iostream>
#include <link.h>
#include <optional>
#include <span>
#include <sstream>
#include <string.h>
#include <sys/mman.h>
#include <unistd.h>
#include <vector>

static uint16_t parse_u16(const uint8_t *&ptr) {
  uint16_t value = *reinterpret_cast<const uint16_t *>(ptr);
  ptr += sizeof(value);
  return value;
}

static uint32_t parse_u32(const uint8_t *&ptr) {
  uint32_t value = *reinterpret_cast<const uint32_t *>(ptr);
  ptr += sizeof(value);
  return value;
}

static uint64_t parse_u64(const uint8_t *&ptr) {
  uint64_t value = *reinterpret_cast<const uint64_t *>(ptr);
  ptr += sizeof(value);
  return value;
}

static Stackmap::Header parse_header(const uint8_t *&ptr) {
  Stackmap::Header header = *reinterpret_cast<const Stackmap::Header *>(ptr);
  ptr += sizeof(header);
  return header;
}

static Stackmap::StkSizeRecord parse_stk_size_record(const uint8_t *&ptr) {
  Stackmap::StkSizeRecord record =
      *reinterpret_cast<const Stackmap::StkSizeRecord *>(ptr);
  ptr += sizeof(record);
  return record;
}

static Stackmap::Constant parse_constant(const uint8_t *&ptr) {
  Stackmap::Constant constant =
      *reinterpret_cast<const Stackmap::Constant *>(ptr);
  ptr += sizeof(constant);
  return constant;
}

static Stackmap::Location parse_location(const uint8_t *&ptr) {
  Stackmap::Location location =
      *reinterpret_cast<const Stackmap::Location *>(ptr);
  ptr += sizeof(location);
  return location;
}

static Stackmap::LiveOut parse_live_out(const uint8_t *&ptr) {
  Stackmap::LiveOut live_out =
      *reinterpret_cast<const Stackmap::LiveOut *>(ptr);
  ptr += sizeof(live_out);
  return live_out;
}

static Stackmap::StkMapRecord parse_stk_map_record(const uint8_t *&ptr) {
  uint64_t patchpoint_id = parse_u64(ptr);
  uint32_t instruction_offset = parse_u32(ptr);
  uint16_t record_flags = parse_u16(ptr);
  uint16_t num_locations = parse_u16(ptr);

  std::vector<Stackmap::Location> locations;
  for (uint16_t i = 0; i < num_locations; i++)
    locations.push_back(parse_location(ptr));

  uint32_t padding1 = 0;
  if ((uint64_t)ptr % 8 != 0) {
    ASSERT((uint64_t)ptr % 8 == 4 && "Invalid data alignment");
    padding1 = parse_u32(ptr);
  }
  uint16_t padding2 = parse_u16(ptr);
  uint16_t num_live_outs = parse_u16(ptr);
  std::vector<Stackmap::LiveOut> live_outs;
  for (uint16_t i = 0; i < num_live_outs; i++)
    live_outs.push_back(parse_live_out(ptr));

  uint32_t padding3 = 0;
  if ((uint64_t)ptr % 8 != 0) {
    ASSERT((uint64_t)ptr % 8 == 4 && "Invalid data alignment");
    padding3 = parse_u32(ptr);
  }

  return Stackmap::StkMapRecord{.patchpoint_id = patchpoint_id,
                                .instruction_offset = instruction_offset,
                                .record_flags = record_flags,
                                .num_locations = num_locations,
                                .locations = locations,
                                .padding1 = padding1,
                                .padding2 = padding2,
                                .num_live_outs = num_live_outs,
                                .live_outs = live_outs,
                                .padding3 = padding3};
}

Stackmap::Stackmap parse_stackmap(std::span<const uint8_t> data) {
  const uint8_t *ptr = data.data();
  ASSERT(ptr != nullptr && "Invalid data");
  ASSERT((uint64_t)ptr % 8 == 0 && "Invalid data alignment");
  Stackmap::Header header = parse_header(ptr);

  uint32_t num_functions = parse_u32(ptr);
  uint32_t num_constants = parse_u32(ptr);
  uint32_t num_records = parse_u32(ptr);

  std::vector<Stackmap::StkSizeRecord> stksize_records;
  for (uint32_t i = 0; i < num_functions; i++)
    stksize_records.push_back(parse_stk_size_record(ptr));

  std::vector<Stackmap::Constant> constants;
  for (uint32_t i = 0; i < num_constants; i++)
    constants.push_back(parse_constant(ptr));

  std::vector<Stackmap::StkMapRecord> stkmap_records;
  for (uint32_t i = 0; i < num_records; i++)
    stkmap_records.push_back(parse_stk_map_record(ptr));

  return Stackmap::Stackmap{.header = header,
                            .num_functions = num_functions,
                            .num_constants = num_constants,
                            .num_records = num_records,
                            .stksize_records = stksize_records,
                            .constants = constants,
                            .stkmap_records = stkmap_records};
}

static std::string location_kind_to_string(Stackmap::LocationKind kind) {
  switch (kind) {
  case Stackmap::LocationKind::REGISTER:
    return "Register";
  case Stackmap::LocationKind::DIRECT:
    return "Direct";
  case Stackmap::LocationKind::INDIRECT:
    return "Indirect";
  case Stackmap::LocationKind::CONSTANT:
    return "Constant";
  case Stackmap::LocationKind::CONSTANT_INDEX:
    return "Constant index";
  default:
    return "Unknown";
  }
}

std::string stackmap_to_string(const Stackmap::Stackmap &stackmap) {
  std::stringstream ss;

  ss << "Version: " << stackmap.header.version << std::endl;
  ss << "Num functions: " << stackmap.num_functions << std::endl;
  ss << "Num constants: " << stackmap.num_constants << std::endl;
  ss << "Num records: " << stackmap.num_records << std::endl;

  for (size_t i = 0; i < stackmap.stksize_records.size(); i++) {
    const Stackmap::StkSizeRecord &record = stackmap.stksize_records[i];
    ss << "StkSizeRecord[" << i << "]" << std::endl;
    ss << "  Address: 0x" << std::hex << record.function_address << std::endl;
    ss << "  Stack size: " << std::dec << record.stack_size << std::endl;
    ss << "  Record count: " << record.record_count << std::endl;
  }

  /*
  for (size_t i = 0; i < stackmap.constants.size (); i++)
    {
      const Stackmap::Constant &constant = stackmap.constants[i];
      ss << "Constant[" << i << "]" << std::endl;
      ss << "  Value: " << constant.large_constant << std::endl;
    }
    */

  for (size_t i = 0; i < stackmap.stkmap_records.size(); i++) {
    const Stackmap::StkMapRecord &record = stackmap.stkmap_records[i];
    ss << "StkMapRecord[" << i << "]" << std::endl;
    ss << "  Patchpoint ID: " << record.patchpoint_id << std::endl;
    ss << "  Instruction offset: " << record.instruction_offset << std::endl;
    ss << "  Record flags: " << record.record_flags << std::endl;
    ss << "  Num locations: " << record.num_locations << std::endl;

    for (size_t j = 0; j < record.locations.size(); j++) {
      const Stackmap::Location &location = record.locations[j];
      ss << "  Location[" << j << "] = ";
      Register reg{location.dwarf_regnum};
      switch (location.kind) {
      case Stackmap::LocationKind::REGISTER: {
        ss << reg_to_string(reg) << std::endl;
      } break;
      case Stackmap::LocationKind::DIRECT: {
        ss << reg_to_string(reg) << " + " << location.offset << std::endl;
      } break;
      case Stackmap::LocationKind::INDIRECT: {
        ss << "[" << reg_to_string(reg) << " + " << location.offset << "]"
           << std::endl;
      } break;
      case Stackmap::LocationKind::CONSTANT: {
        ss << location.offset << std::endl;
      } break;
      case Stackmap::LocationKind::CONSTANT_INDEX: {
        ASSERT(location.offset < stackmap.constants.size() &&
               "Invalid constant index");
        ss << "Constants[" << location.offset
           << "] = " << stackmap.constants[location.offset].large_constant
           << std::endl;
      } break;
      default:
        ss << "Unknown" << std::endl;
        break;
      }
    }
  }

  return ss.str();
}
